[docs][Autocomplete] Fix GoogleMaps demo

hbjORbj opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Closes https://github.com/mui/material-ui/issues/35391

**Problem**:
- For some queries, the [Google Maps Autocomplete demo](https://mui.com/material-ui/react-autocomplete/#GoogleMaps.js) crashes

**Before**:
(1) Go to https://mui.com/material-ui/react-autocomplete/#google-maps-place
(2) Enter "Darse de Baudour"
(3) Demo crashes

**After**: 
(1) Go to https://deploy-preview-35545--material-ui.netlify.app/material-ui/react-autocomplete/#google-maps-place
(2) Enter "Darse de Baudour"
(3) Demo doesn't crash

oliviertassinari commented on docs/data/material/components/autocomplete/GoogleMaps.tsx:
```diff
@@ -138,6 +139,13 @@ export default function GoogleMaps() {
       )}
       renderOption={(props, option) => {
         const matches = option.structured_formatting.main_text_matched_substrings;
+        if (!matches) {
+          return (
+            <ListItem sx={{ color: 'text.secondary', pt: '6px', pb: '6px' }}>
+              No options
+            </ListItem>
+          );
+        }
```
I don't understand this logic. If there is an option, there are no "No options". So, if we want to handle no matching substring, I think that we should do it before.

No, Google Maps returns an option, ok, there is no matched substring, but it could still be what the person meant?

<img width="409" alt="Screenshot 2022-12-21 at 01 13 09" src="https://user-images.githubusercontent.com/3165635/208797479-f579bba9-a372-438c-8464-90c4d916ba09.png">

hbjORbj replied:
Thanks, addressed.

oliviertassinari reviewed (commented):
A few more ideas to improve the Google Maps place demo:

1. There is an overflow issue, e.g. when searching for "zzzz"

<img width="324" alt="Screenshot 2022-12-21 at 01 16 21" src="https://user-images.githubusercontent.com/3165635/208797654-18ed8f8a-f2a3-4f23-b710-a4be39319676.png">

2. Should we customize the `noOptionsText` prop? Maybe like "No places" would make more sense than "No options"?
3. The warning "Before you can start using the Google Maps JavaScript API and Places API, you must sign up and create a billing account." feels unclear. It doesn't explain why. I think that it should mention that it's to get an [API key](https://developers.google.com/maps/documentation/javascript/get-api-key).
4. We frequently hit a rate limit error. I have configured some of it in our [Google Cloud dashboard](https://console.cloud.google.com/google/maps-apis/quotas?project=material-ui-docs&pageState=(%22duration%22:(%22groupValue%22:%22P14D%22,%22customValue%22:null))) so we stay inside the free tier, but I don't understand why we get them. Maybe we could set the throttle at 400ms rather than 200ms?

  <img width="468" alt="Screenshot 2022-12-21 at 01 37 19" src="https://user-images.githubusercontent.com/3165635/208800034-0ddc554d-b21d-4948-beed-4b632381fc66.png">

5. Considering how we solve the problem of out-of-order requests, using throttle makes no sense. We effectively have a debounce, so we might as well use a debounce to make fewer network requests.

hbjORbj commented:
> 1. There is an overflow issue, e.g. when searching for "zzzz"

Fixed.

> 2. Should we customize the noOptionsText prop? Maybe like "No places" would make more sense than "No options"?

Agree. I will go with `"No locations"`.

> 3. The warning "Before you can start using the Google Maps JavaScript API and Places API, you must sign up and create a billing account." feels unclear. It doesn't explain why. I think that it should mention that it's to get an [API key](https://developers.google.com/maps/documentation/javascript/get-api-key).

Yes, "Before you can start using the Google Maps JavaScript API and Places API, you need to get your own [API key](https://developers.google.com/maps/documentation/javascript/get-api-key)." sounds better.

hbjORbj commented:
> 4. We frequently hit a rate limit error. I have configured some of it in our [Google Cloud dashboard](https://console.cloud.google.com/google/maps-apis/quotas?project=material-ui-docs&pageState=(%22duration%22:(%22groupValue%22:%22P14D%22,%22customValue%22:null))) so we stay inside the free tier, but I don't understand why we get them. Maybe we could set the throttle at 400ms rather than 200ms?
> 5. Considering how we solve the problem of out-of-order requests, using throttle makes no sense. We effectively have a debounce, so we might as well use a debounce to make fewer network requests.

Replaced `throttle` with `debounce` from `'@mui/material/utils'` and changed 200ms to 400ms in this [commit](7a6112e30f07cccb780daaa645330f70e3ba12f8)

oliviertassinari reviewed (commented):
I have pushed a few more changes that were not directly related to the initial problem but seemed to be a good opportunity to modernize the example, to use the features that were not available back then.

I can't find anything else to improve 👍

oliviertassinari commented:
**Off-topic**. I have just realized that the undo <kbd>Ctrl</kbd>+<kbd>Z</kbd> and redo shortcuts do nothing on our autocomplete. It's a bit frustrating. It works in Gmail:

<img width="332" alt="Screenshot 2022-12-21 at 21 28 00" src="https://user-images.githubusercontent.com/3165635/209005697-02e993ce-7d36-4624-9135-04f4feef3fdc.png">

I can't find any UI component that supports this (select2, react-select, downshift, react-spectrum, they don't), so maybe an opportunity to get an edge on the UX side.

hbjORbj commented:
> I can't find any UI component that supports this (select2, react-select, downshift, react-spectrum, antd, they don't), so maybe an opportunity to get an edge on the UX side. Or maybe it's overkill 🤷‍♂️

I don't really think it's an overkill ;) It's quite important for UX that undo/redo works.

mnajdova reviewed (approved):
No way this really happened (I wanted to test the "No locations" scenario) :D 

![image](https://user-images.githubusercontent.com/4512430/209158881-c769739e-b7cf-486d-b03d-0a4fbe2042ef.png)

The updates look good 👍
